Property flags and descriptors (#196)
javascript-tutorial/it.javascript.info: 1-js/07-object-properties/01-property-descriptors/article.md

# Attributi e descrittori di proprietà

> Come già sappiamo, gli oggetti possono memorizzare proprietà

Reviewer: ```suggestion
Come già sappiamo, gli oggetti possono memorizzare proprietà.
```

Fino ad ora, per noi, una proprietà è sempre stata una coppia "chiave-valore". Ma in realtà, una proprietà è molto più potente e flessibile di cosi.

In questo capitolo studieremo ulteriori opzioni di configurazione, e nel prossimo vedremo come trasformarle in funzioni getter/setter.

## Attributi di proprietà

Le proprietà degli oggetti, oltre ad un **`valore`**, possiedono tre attributi speciali (cosi detti "flags", o "bandiere"):

- **`writable`** -- se impostato a `true`, il valore può essere modificato, altrimenti è possibile accedervi in sola lettura.

> - **`enumerable`** -- se impostato a `true`, potrà apparire nei loop, altrimenti non verrà considerata.

Reviewer: ```suggestion
- **`enumerable`** -- se impostato a `true`, appare nei loop, altrimenti non verrà considerata.
```
Mi sembra piu' chiaro e semplice cosi', che ne pensi?

> - **`configurable`** -- se impostato a `true`, la proprietà la proprietà può essere cancellata e questi attributi possono essere modificati.

Reviewer: ```suggestion
- **`configurable`** -- se impostato a `true`, la proprietà può essere cancellata e questi attributi possono essere modificati.
```

Non li abbiamo mai visti fino ad ora, perché generalmente non vengono mostrati. Quando creiamo una proprietà in "modo ordinario", questi attributi vengono tutti impostati a `true`. Ma possiamo comunque modificarli in qualsiasi momento.

Come prima cosa, vediamo come poter accedere a questi attributi.

Il metodo [Object.getOwnPropertyDescriptor](mdn:js/Object/getOwnPropertyDescriptor) ritorna *tutte* le informazioni riguardo una proprietà.

La sintassi:
```js
let descriptor = Object.getOwnPropertyDescriptor(obj, propertyName);
```

`obj`

> : L'oggetto da cui volgiamo ottenere le informazioni.

Reviewer: ```suggestion
: L'oggetto da cui vogliamo ottenere le informazioni.
```
(e quei `:` a che servono?)

*/
```

Per modificare gli attributi possiamo utilizzare [Object.defineProperty](mdn:js/Object/defineProperty).

La sintassi:

```js
Object.defineProperty(obj, propertyName, descriptor)
```

`obj`, `propertyName`

> : L'oggetto e la proprietà in cui applicare il descrittore.

Reviewer: ```suggestion
: L'oggetto e la proprietà a cui applicare il descrittore.
```
Oppure:

```suggestion
: L'oggetto e la proprietà cui applicare il descrittore.
```
*In cui* non e' corretto

`descriptor`

> : Descrittore di proprietà dell'oggetto da applicare.

Reviewer: ```suggestion
: Descrittore di proprietà da utilizzare.
```
*applicare* non mi convince tanto, nemmeno nella versione inglese. In questo contesto non si capisce bene, sembra anzi che quel che dobbiamo applicare sia *l'oggetto*. Anche *dell'oggetto*; piu' sopra andava anche bene trovo, ma qui non ha lo stesso significato che ha in inglese; ovvero, nella versione inglese si capisce che *property descriptor* e' un tipo di oggetto, mentre noi ci riferiamo a questo oggetto come a un *descrittore di proprieta' ***dell***'oggetto*. Che ne dici di usare 'un oggetto descrittore di proprieta' ' ?

> Se la proprietà esiste, `defineProperty` aggiornerà l'attributo. Altrimenti, creerà la proprietà con il valore e gli attributi forniti, se un attributo non viene fornito, gli verrà assegnato il valore `false`.

Reviewer: ```suggestion
Se la proprietà esiste, `defineProperty` aggiornerà l'attributo. Altrimenti, creerà la proprietà con il valore e gli attributi forniti; se un attributo non viene fornito, gli verrà assegnato il valore `false`.
```
piccolo suggerimento, la frase e' piu' facile da leggere cosi' :)

> Ad esempio. qui creiamo una proprietà `name` con tutti gli attributi a `false`:

Reviewer: ```suggestion
Ad esempio, qui creiamo una proprietà `name` con tutti gli attributi `false`:
```
Quel *a* e' superfluo

 */
```

> Confrontandola con la proprietà "creata normalmente" `user.name` vista sopra, ora tutti gli attributi sono a `false`. Se questo non è ciò che vogliamo, allora dovremmo impostarli a `true` tramite il `descriptor`.

Reviewer: ```suggestion
Confrontandola con la proprietà "creata normalmente" `user.name` vista sopra, ora tutti gli attributi sono `false`. Se questo non è ciò che vogliamo, allora dovremmo impostarli a `true` tramite il `descriptor`.
```
Quando dici impostarli *a* va bene; dire "sono *a*" non va bene

> Ora analizziamo gli effetti degli attributi guardando ad alcuni esempi.

Reviewer: ```suggestion
Ora analizziamo gli effetti degli attributi guardando alcuni esempi.
```

## Non-writable

> Vediamo come rendere `user.name` non-writable (non può essere riassegnata), modificando l'attributo `writable`:

Reviewer: ```suggestion
Vediamo come rendere `user.name` *non-writable* (la variabile non può essere riassegnata) modificando l'attributo `writable`:
```
 Piccola pignoleria: ho aggiunto *la variabile* perche' ora ci riferiamo a `user.name`, maschile, con riassegnata, femminile. E' una piccolezza, ma la piccola aggiunta chiarisce che non ci stiamo definendo a `user.name` ma a cio' che rappresenta, cioe' una variabile

> */!*

Reviewer 2: Ci sarebbe il commento nel codice da tradurre, linea 119

Author: Qui non ho tradotto perché suppongo (ma non sono sicuro) che sia il messaggio di errore che ti appare.. e ho paura che traducendolo perda di senso, visto che poi non apparirà mai in italiano... però non ci metterei la mano sul fuoco, smentitemi se sto dicendo una cavolata

```

Ora nessuno potrà modificare il nome dell'utente, a meno che non vada a sovrascrivere il valore degli attributi con `defineProperty`.

```smart header="Gli errori verranno mostrati solamente in strict mode"

> Se non siamo in "strict mode", e tentiamo di sovrascrivere una proprietà non-writable, non verrà mostrato alcun errore. Nonostante non venga mostrato l'errore, l'operazione fallirà comunque. Quindi le violazioni di attributi fuori dalla strict mode verranno ignorate silenziosamente.

Reviewer: ```suggestion
Se non siamo in "strict mode", e tentiamo di sovrascrivere una proprietà non-writable, non verrà mostrato alcun errore. Nonostante non venga mostrato l'errore, l'operazione fallirà comunque. Quindi le violazioni di attributi fuori dalla strict mode verranno ignorate.
```

Quel *silenziosamente* suona un po' male, non funziona come nella versione inglese. Che ne pensi? Al massimo inverti i due:

```suggestion
Se non siamo in "strict mode", e tentiamo di sovrascrivere una proprietà non-writable, non verrà mostrato alcun errore. Nonostante non venga mostrato l'errore, l'operazione fallirà comunque. Quindi le violazioni di attributi fuori dalla strict mode verranno silenziosamente  ignorate.
```

```

> Qui vediamo lo stesso esempio, ma la proprietà viene creata da zero:

Reviewer: ```suggestion
Qui vediamo lo stesso esempio, ma la proprietà viene creata dal nulla:
```

```js run
let user = { };

Object.defineProperty(user, "name", {
*!*
  value: "John",

>   // per le nuove proprietà dobbiamo esplicitare quali attributi valgono true

Reviewer: ```suggestion
  // per le nuove proprietà dobbiamo esplicitare quali attributi sono true
```

## Non-enumerable

Ora proviamo ad aggiungere un metodo `toString` ad `user`.

> Normalmente, la funzione built-in (integrata) `toString` per gli oggetti, è non-enumerable, quindi non verrà mostrata nei cicli come `for..in`. Ma se proviamo ad aggiungere una nostra definizione di `toString`, allora questa verrà mostrata nei cicli `for..in`, come nell'esempio:

Reviewer: ```suggestion
Normalmente, la funzione *built-in* (integrata) `toString` , per gli oggetti è non-enumerable, quindi non verrà mostrata nei cicli come `for..in`. Ma se proviamo ad aggiungere una nostra definizione di `toString`, allora questa verrà mostrata nei cicli `for..in`, come nell'esempio:
```

  }
};

// Di default, entrambe le proprietà verranno elencate
for (let key in user) alert(key); // name, toString
```

> Se non è cio che ci aspettiamo, possiamo impostare l'attributo `enumerable:false`. In questo modo non verrà più mostrata nei cicli `for..in`, propriò come la funzione integrata definita da JavaScript:

Reviewer: ```suggestion
Se non è ciò che ci aspettiamo, possiamo impostare l'attributo `enumerable:false`. In questo modo non verrà più mostrata nei cicli `for..in`, proprio come la funzione  già integrata (definita da Javascript):
```

## Non-configurable

> L'attributo non-configurable (`configurable:false`), talvolta è preimpostato negli gli oggetti e nelle proprietà integrate.

Reviewer: ```suggestion
L'attributo non-configurable (`configurable:false`) è talvolta preimpostato negli oggetti e nelle proprietà integrate.
```
(Un *built-in* qui non sarebbe male, al posto di 'integrate'; fuori contesto, senza un esempio come quello sopra con `toString`, non si capisce bene che cosa sia integrato, e dove)

> Una proprietà non-configurable non può essere cancellata.

Reviewer: ```suggestion
Una proprietà *non-configurable* non può essere cancellata.
```

> Ad esempio, `Math.PI` è non-writable, non-enumerable e non-configurable:

Reviewer: ```suggestion
Ad esempio, `Math.PI` è *non-writable*, *non-enumerable* e *non-configurable*:
```

}
*/
```

> Quindi, uno sviluppatore, non sarà in grado di cambiare il valore `Math.PI` o sovrascriverlo.

Reviewer: ```suggestion
Quindi, uno sviluppatore non sarà in grado di cambiare il valore `Math.PI` o di sovrascriverlo.
```

```js run
Math.PI = 3; // Error

> // non si potrebbe comunque cancellare Math.PI

Reviewer 2: ```suggestion
// e nemmeno la cancellazione Math.PI funzionerebbe
```

```

> Rendere una proprietà non-configurable è una "strada a senso unico". Non possiamo tornare indietro con `defineProperty`.

Reviewer: ```suggestion
Rendere una proprietà *non-configurable* è una "strada a senso unico". Non possiamo tornare indietro tramite `defineProperty`.
```

> Per essere precisi, l'attributo non-configurable impone diverse restrizioni su `defineProperty`:

Reviewer: ```suggestion
Per essere precisi, l'attributo non-configurable impone diverse restrizioni a `defineProperty`:
```

1. Non possiamo modificare l'attributo `configurable`.
2. Non possiamo modificare l'attributo `enumerable`.

> 3. Non possiamo modificare l'attributo da `writable: false` a `true` (possiamo invece modificarla da `true` a `false`).

Reviewer: ```suggestion
3. Non possiamo modificare l'attributo da `writable: false` a `true` (possiamo invece modificarlo da `true` a `false`).
```
Attributo e' maschile

4. Non possiamo modificare le funzioni di accesso `get/set` (ma possiamo assegnarle nel caso non siano definite).

**L'idea alla base di "configurable: false" è quella di prevenire la modifica e la rimozione degli attributi di una proprietà, permettendo comunque la modifica del suo valore.**

> In questo esempio `user.name` è non-configurable, ma possiamo comunque modificarlo (poiché è writable):

Reviewer: ```suggestion
In questo esempio `user.name` è *non-configurable*, ma possiamo comunque modificarlo (poiché è *writable*):
```

});
```

In questo modo siamo in grado di impostare più proprietà in una volta sola.

## Object.getOwnPropertyDescriptors

Per ottenere tutti i descrittori di una proprietà, possiamo utilizzare il metodo [Object.getOwnPropertyDescriptors(obj)](mdn:js/Object/getOwnPropertyDescriptors).

Il metodo `Object.defineProperties`  può essere utilizzato per clonare un oggetto mantenendo gli attributi delle sue proprietà:

```js
let clone = Object.defineProperties({}, Object.getOwnPropertyDescriptors(obj));
```

> Normalmente quando cloniamo un oggetto, utilizziamo l'assegnazione per copiarne le proprietà, come nell'esempio:

Reviewer: ```suggestion
Normalmente, quando cloniamo un oggetto, utilizziamo l'assegnazione per copiarne le proprietà, come nell'esempio:
```

```js
for (let key in user) {
  clone[key] = user[key]
}
```

> ...Ma in questo modo non stiamo copiando gli attributi. Quindi per una clonazione più completa, l'utilizzo di `Object.defineProperties` è la strada migliore.

Reviewer: ```suggestion
...Ma in questo modo non stiamo copiando gli attributi. Quindi per una clonazione più completa, l'utilizzo di `Object.defineProperties` è la scelta migliore.
```

> Un'altra differenza è che `for..in` ignora le proprietà di tipo symbol, invece `Object.getOwnPropertyDescriptors` ritorna *tutti* i descrittori, inclusi quelli di tipi symbol.

Reviewer: ```suggestion
Un'altra differenza è che `for..in` ignora le proprietà di tipo `symbol`, mentre `Object.getOwnPropertyDescriptors` ritorna *tutti* i descrittori, inclusi quelli di tipo symbol.
```

## Sigillare un oggetto globalmente

I descrittori di proprietà permettono di lavorare a livello di proprietà.

Esistono però diversi metodi in grado di limitare l'accesso *all'intero* oggetto:

[Object.preventExtensions(obj)](mdn:js/Object/preventExtensions)
: Vieta di aggiungere nuove proprietà all'oggetto.

[Object.seal(obj)](mdn:js/Object/seal)
: Vieta di aggiungere/rimuovere proprietà, ed imposta `configurable: false` su tutte le proprietà già esistenti dell'oggetto.

[Object.freeze(obj)](mdn:js/Object/freeze)
: Vieta di aggiungere/rimuovere/modificare le proprietà dell'oggetto. Imposta `configurable: false, writable: false` su tutte le proprietà già esistenti dell'oggetto.

> Ed esistono anche dei metodi verificare lo stato di un oggetto:

Reviewer: ```suggestion
Ed esistono anche dei metodi per verificare lo stato degli attributi di un oggetto:
```

[Object.isExtensible(obj)](mdn:js/Object/isExtensible)
: Ritorna `false` se è vietato aggiungere nuove proprietà, altrimenti ritorna `true`.

[Object.isSealed(obj)](mdn:js/Object/isSealed)
: Ritorna `true` se è vietato aggiungere/rimuovere proprietà, e tutte le altre proprietà sono impostate a `configurable: false`.

[Object.isFrozen(obj)](mdn:js/Object/isFrozen)
: Ritorna `true` se è vietato aggiungere/rimuovere/modificare proprietà, e tutte le altre proprietà sono impostate a `configurable: false, writable: false`.

> Nella pratica questi metodi sono utilizzati molto raramente.

Reviewer: ```suggestion
In pratica, tuttavia, questi metodi sono utilizzati molto raramente.
```
